iocore/hostdb/P_RefCountCache.h

```diff
@@ -0,0 +1,896 @@
+            this->version.ink_minor == other.version.ink_minor &&
+            this->object_version.ink_major == other.object_version.ink_major &&
+            this->object_version.ink_minor == other.object_version.ink_minor);
+  }
+
+  bool
+  compatible(RefCountCacheHeader *other)
+  {
+    return (this->magic == other->magic && this->version.ink_major == other->version.ink_major &&
+            this->object_version.ink_major == other->version.ink_major);
+  };
+};
+
+// This continuation is responsible for persisting RefCountCache to disk
+// To avoid locking the partitions for a long time we'll do the following per-partition:
+//    - lock
+//    - copy ptrs (bump refcount)
+//    - unlock
+//    - persist
+//    - remove ptrs (drop refcount)
+// This way we only have to hold the lock on the partition for the time it takes to get Ptr<>s to all items in the partition
+template <class C> class RefCountCacheSync : public Continuation
+{
+public:
+  typedef int (RefCountCacheSync::*CacheSyncHandler)(int, void *);
+
+  size_t partition; // Current partition
+  C *cc;            // Pointer to the entire cache
+  Continuation *cont;
+
+  int copy_partition(int event, Event *e);
+  int write_partition(int event, Event *e);
+  int pause_event(int event, Event *e);
+
+  // Create the tmp file on disk we'll be writing to
+  int initialize_storage(int event, Event *e);
+  // do the final mv and close of file handle
+  int finalize_sync();
+
+  // helper method to spin on writes to disk
+  int write_to_disk(char *i, int size);
+
+  RefCountCacheSync(Continuation *acont, C *cc, int frequency, std::string dirname, std::string filename);
+
+private:
+  Vec<RefCountCacheHashEntry *> partition_items;
+
+  int fd; // fd for the file we are writing to
+
+  std::string dirname;
+  std::string filename;
+  std::string tmp_filename;
+
+  ink_hrtime time_per_partition;
+  ink_hrtime start;
+
+  int total_items;
+  int64_t total_size;
+
+  RecRawStatBlock *rsb;
+
+  SocketManager socket_manager;
+};
+
+template <class C>
+int
+RefCountCacheSync<C>::copy_partition(int event, Event *e)
+{
+  (void)event;
+  if (partition >= cc->partition_count()) {
+    int sync_ret = this->finalize_sync();
+    if (sync_ret != 0) {
+      Warning("Unable to finalize sync of cache to disk %s: %d", this->filename.c_str(), sync_ret);
+    }
+    cont->handleEvent(REFCOUNT_CACHE_EVENT_SYNC, 0);
+    Debug("refcountcache", "RefCountCacheSync done");
+    delete this;
+    return EVENT_DONE;
+  }
+  Debug("refcountcache", "sync partition=%ld/%ld", partition, cc->partition_count());
+  // copy the partition into our buffer, then we'll let `pauseEvent` write it out
+  this->partition_items.reserve(cc->get_partition(partition).count());
+  cc->get_partition(partition).copy(this->partition_items);
+  partition++;
+  SET_HANDLER((CacheSyncHandler)&RefCountCacheSync::write_partition);
+  mutex = e->ethread->mutex;
+  e->schedule_imm(ET_TASK);
```

Comment:
Yeh I thought about this some more. Plugin metrics don't work on ``ET_TASK`` threads because they don't have a stats block. Since you are passing the stats block down, I guess it should be ok.